src/stan_math_backend/Transform_Mir.ml

```diff
@@ -1,6 +1,64 @@
 open Core_kernel
 open Middle
 
+let use_opencl = ref false
+
+let opencl_triggers =
+  String.Map.of_alist_exn
+    [ ( "normal_id_glm_lpdf"
+      , ( [0; 1]
+        , [ (* Array of conditions under which to move to OpenCL *) ([1], [])
+          (* Argument 1 is data *)
+           ] ) )
+    ; ("bernoulli_logit_glm_lpmf", ([0; 1], [([1], [])]))
+    ; ( "categorical_logit_glm_lpmf"
+      , ([0; 1], [([1], [(1, UMatrix)])])
+        (* Additional requirement of argument 1 being a matrix *) )
+    ; ("neg_binomial_2_log_glm_lpmf", ([0; 1], [([1], [])]))
+    ; ("ordered_logistic_glm_lpmf", ([0; 1], [([1], [(1, UMatrix)])]))
+    ; ("poisson_log_glm_lpmf", ([0; 1], [([1], [])])) ]
+
+let opencl_suffix = "_opencl__"
+let to_matrix_cl e = {e with expr= FunApp (StanLib, "to_matrix_cl", [e])}
+
+let rec switch_expr_to_opencl available_cl_vars e =
+  let is_avail = List.mem available_cl_vars ~equal:( = ) in
+  let to_cl e =
+    match e.expr with
+    | Var s when is_avail s -> {e with expr= Var (s ^ opencl_suffix)}
+    | _ -> to_matrix_cl e
+  in
+  let move_cl_args index arg cl_args =
+    if List.mem ~equal:( = ) cl_args index then to_cl arg else arg
+  in
+  let check_type (i, t) args = (List.nth_exn args i).emeta.mtype = t in
+  let check_if_data ind args =
+    match (List.nth_exn args ind).expr with
+    | Var s when is_avail s -> true
+    | _ -> false
+  in
+  let req_met (data_arg, type_arg) args =
+    List.for_all ~f:(fun x -> check_if_data x args) data_arg
+    && List.for_all ~f:(fun x -> check_type x args) type_arg
+  in
+  let any_req_met args req_args =
+    List.exists ~f:(fun x -> req_met x args) req_args
+  in
+  let get_trigger f = Map.find_exn opencl_triggers f in
+  let first (a, _) = a in
+  let second (_, a) = a in
+  let cl_args f = first (get_trigger f) in
+  let req_args f = second (get_trigger f) in
+  let map_args args f =
+    if any_req_met args (req_args f) then
```

Comment:
I am almost certain there is an easier way to get (req_args f), same for cl_args.